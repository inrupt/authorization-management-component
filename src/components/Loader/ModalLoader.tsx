//
// Copyright Inrupt Inc.
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal in
// the Software without restriction, including without limitation the rights to use,
// copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
// Software, and to permit persons to whom the Software is furnished to do so,
// subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
// INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
// PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
// HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
// OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
// SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
//
import { useEffect, useState } from "react";
import { Modal } from "reactstrap";
import Loader from "./Loader";
import styles from "./Loader.module.scss";

/**
 * Used to block interaction with the screen and display a loading symbol
 */
export default function ModalLoader({ isOpen }: { isOpen: boolean }) {
  const [loaderTimer, setLoaderTimer] = useState<{
    timeout?: ReturnType<typeof setTimeout> | undefined;
    loader: boolean;
  }>({ loader: false });

  useEffect(() => {
    if (isOpen) {
      setLoaderTimer({
        timeout: setTimeout(() => {
          setLoaderTimer({ loader: true });
        }, 1_500),
        loader: false,
      });
    } else {
      clearTimeout(loaderTimer.timeout);
      setLoaderTimer({ loader: false });
    }
    // The linter wants loaderTimer.timeout to be a dependency here.
    // This is not necessary.
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [isOpen]);

  return (
    <Modal
      className={styles["custom-modal"]}
      isOpen={isOpen}
      contentClassName={styles["custom-modal-content"]}
      title="loading screen"
      aria-label="loading screen"
      labelledBy="loading-div"
    >
      <Loader hidden={!loaderTimer.loader} />
    </Modal>
  );
}
