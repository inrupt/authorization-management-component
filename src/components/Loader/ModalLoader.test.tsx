// MIT License
//
// Copyright Inrupt Inc.
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in all
// copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.
//

import { render, waitFor } from "@testing-library/react";
import * as axe from "axe-core";
import ModalLoader from "./ModalLoader";

describe("ModalLoader", () => {
  it("renders a full page loader", async () => {
    const { asFragment, container, getByTestId } = render(
      <ModalLoader isOpen />,
    );

    await waitFor(
      () => {
        expect(getByTestId("loading-spinner")).toBeVisible();
      },
      {
        timeout: 2_000,
      },
    );

    expect(asFragment()).toMatchSnapshot();

    // Accessibility tests
    HTMLCanvasElement.prototype.getContext = jest.fn();
    const results = await axe.run(container);
    expect(results.violations).toHaveLength(0);
  });
});
