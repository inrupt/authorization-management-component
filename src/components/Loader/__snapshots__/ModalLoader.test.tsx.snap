// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ModalLoader renders a full page loader 1`] = `<DocumentFragment />`;
